[needs-docs] Show 'template layer' constraints info in the Refactor fields algorithm's UI

gacarrillor opened the pull request:
## Description
Dialog: Refactor fields algorithm's UI.
Objective: Highlight field rows when the 'template layer' field has constraints.

**TODO**
 + [x] Add an orange color (same as used in attribute form warnings) to the whole row if a template layer field has constraints.
 + [x] Adjust text in help panel to let users know why some rows are coloured. 
 + [x] Remove NOT NULL column.
 + [x] New column "Template properties."
     + Which content to show?
        + "Constraints active"?
     + What the tooltip should show?
        + Constraints, widget configuration, ...?
 + [x] Replace "Load fields from layer" by "Load fields from template layer."

**Screenshots**

![image](https://user-images.githubusercontent.com/652785/75100644-6079a500-559e-11ea-91b6-f99fa05451de.png)


Fix #34458




## Checklist

<!-- Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.
-->

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment.

gacarrillor committed:
Highlight NOT NULL target fields in the Refactor fields algorithm's UI:

+ By adding a new 'NOT NULL' column in the mapping widget, and
+ By adding an orange color (same as used in attribute form warnings) to the whole row.

New paragraph added to help panel to let users know why some rows are coloured.

nyalldawson commented on python/plugins/processing/algs/help/qgis.yaml:
```diff
@@ -466,6 +466,8 @@ qgis:refactorfields: >
 
   The original layer is not modified. A new layer is generated, which contains a modified attribute table, according to the provided fields mapping.
 
+  Rows in orange are not allowed to be NULL in the layer used as target. However, since the algorithm creates a new layer with no constraints, this can be taken only as a hint.
```
I really don't see the value of this given the listed limitation. It's very confusing for users, who would expect this flag to apply to the newly created layer.

gacarrillor replied:
@nyalldawson, if you use the Refactor Fields algorithm, you know it's a bit overwhelming, this PR aims to make it easier for users to prioritize field mappings. 

We actually use the refactor fields along with an "Append Features to Layer" algorithm (3rd party plugin that I also suggested to port to the QGIS core) to send the result into an existing layer. So, it acts as an entire ETL algorithm where knowing the target field constraints makes (a lot of) sense, especially for layers with a large number of fields.

nyalldawson replied:
I still don't see the value sorry, and foresee the inevitable bug reports incoming about how the value isn't reflected in the output layer. Maybe for your particular use case you are best off subclassing the existing algorithm into your plugin so that it can be designed to meet your particular use case, rather then changing the core one itself.

arnaud-morvan replied:
I agree with @nyalldawson that this is very confusing.
This only give the information that the layer used as a template for fields have not null constraints, but the purpose of this algorithm is to create a new layer.
This would make sense in an algorithm that writes features in an existing layer.

gacarrillor commented:
Understood, closing...

m-kuhn commented:
Reopened for reconsideration, as this has been seen as a wider requirement (i.e. this is not an isolated use case and therefore a generic approach as proposed in here is preferable over a subclassing approach). As stated in the below ticket, if this is considered as too intrusive, the coloring can be made optional by a switch / checkbox.

https://github.com/qgis/QGIS/issues/34458

m-kuhn commented on python/plugins/processing/algs/help/qgis.yaml:
```diff
@@ -466,6 +466,8 @@ qgis:refactorfields: >
 
   The original layer is not modified. A new layer is generated, which contains a modified attribute table, according to the provided fields mapping.
 
+  Rows in orange are not allowed to be NULL in the layer used as target. However, since the algorithm creates a new layer with no constraints, this can be taken only as a hint.
```
```suggestion
  Rows in orange have a NOT NULL constraint in the layer from which these fields were loaded.
  Treat this information as a hint during configuration. No constraints will be added on an output layer nor will this be checked or enforced by the algorithm.
```

I think that's slightly more on topic.

gacarrillor replied:
Added a very similar text in the end, but now not only limited to NOT NULL constraints.

nyalldawson commented:
Honestly, I'm still -1. If this is crucial functionality, I think it needs to be exposed in a much less confusing way. My feeling is always that if **I'm** (as a qgis developer) struggling to understand an option in qgis, a regular user will find that option incredibly confusing...

m-kuhn commented:
It is very helpful. If you have an existing target layer (or template thereof), showing it's constraints is very helpful. Especially since the modeler dialog is modal.

I think the most misleading part is the "NOT NULL" column. It looks very much like the other columns which result in real column properties of the created layer. Would it be sufficient to just remove that?

nyalldawson commented:
> It is very helpful. If you have an existing target layer (or template thereof)

Can you clarify what you mean by "target layer"? Currently there's no concept of that at all, you only have a source layer.

m-kuhn commented:
I see the layer specified in "load fields from layer" as a "template" of the target layer.

Situation:
Customer has an old database and a new database schema and wants to migrate data from db version 1 (source) to db version 2 (target). He loads both layers and uses refactor fields to make data match. To configure the algorithm the constraint information of the "template target layer" at hand is a big help.

Currently this means repeatedly closing/reopening this dialog and layer properties to check for constraints (because both dialogs are modal they cannot be opened side by side).

Currently this also means a final manual step to copy/paste all features into the target layer in the end (or use the append to layer plugin as mentioned by @gacarrillor). This could in the future be done automatically but even then will never be within the scope of this algorithm.

nyalldawson commented:
> I see the layer specified in "load fields from layer" as a "template" of the target layer.

Ok fair enough, I'll concede to that :smile:

But can we improve the UI here? I'm still concerned that the original way of exposing this is going to be very confusing for users. Just brainstorming, but what about a new column for "Source Properties" which shows a tooltip with detailed explanatory text when hovered over?

m-kuhn commented:
Agreed on improving the UX.
An additional column with this information would be good. I'd avoid the term "source" though to avoid confusion. It might be good to introduce a new term here. Adding "template" to the load fields from layer label and using template in the column header could help a user link the two.

nyalldawson commented:
> An additional column with this information would be good. I'd avoid the term "source" though to avoid confusion. It might be good to introduce a new term here. Adding "template" to the load fields from layer label and using template in the column header could help a user link the tw

Yes, I like this idea!

gacarrillor commented:
Happy to see how this PR has evolved.
I can continue working on it. I've written the required adjustments down in the PR's description. 
Feel free to suggest changes/adjustments to that list.

gacarrillor committed:
'Template properties' column for Refactor Fields algorithm

as a hint on template/target layer constraints.
A row of a template layer field with constraints is coloured and a tooltip shows the constraints in the corresponding 'Template properties' cell.

gacarrillor committed:
Refactor fields UI: Add 'template' to label to link that layer to new column

gacarrillor committed:
Ensure backwards compatibility with calls not passing 'constraints' key

e.g., processing.execAlgorithmDialog(..., 'FIELDS_MAPPING':[{'expression': 't_id', 'length': -1, 'name': 't_id', 'precision': 0, 'type': 4}], ...)

m-kuhn commented on python/plugins/processing/algs/help/qgis.yaml:
```diff
@@ -330,7 +330,7 @@ qgis:refactorfields: >
   The original layer is not modified. A new layer is generated, which contains a modified attribute table, according to the provided fields mapping.
 
-  Rows in orange are not allowed to be NULL in the layer used as target. However, since the algorithm creates a new layer with no constraints, this can be taken only as a hint.
+  Rows in orange have constraints in the template layer from which these fields were loaded. Treat this information as a hint during configuration. No constraints will be added on an output layer nor will this be checked or enforced by the algorithm.
```
```suggestion
  Rows in orange have constraints in the template layer from which these fields were loaded. Treat this information as a hint during configuration. No constraints will be added on an output layer nor will they be checked or enforced by the algorithm.
```

I think it makes it slightly more readable if it refers to the first part of the sentence.

gacarrillor replied:
Applied!

m-kuhn commented on python/plugins/processing/algs/qgis/ui/FieldsMappingPanel.py:
```diff
@@ -199,7 +213,12 @@ def data(self, index, role=Qt.DisplayRole):
+
+        if role == Qt.ToolTipRole:
+            if column_def['name'] == 'constraints':
+                return ", ".join(
```
```suggestion
                return "<br> ".join(
```

Would this make it more readable?

gacarrillor replied:
I think so. Applied on a single commit with the other suggestion. (This one shows an Outdated tag and it is not possible to commit it from GitHub)

gacarrillor committed:
Improve readability in refactor fields UI

m-kuhn commented:
@nyalldawson how does that look to you?
